internal/weldr/api.go

```diff
@@ -453,13 +453,72 @@ func (api *API) blueprintsDepsolveHandler(writer http.ResponseWriter, request *h
 	})
 }
 
-func (api *API) blueprintsDiffHandler(writer http.ResponseWriter, request *http.Request, _ httprouter.Params) {
-	var reply struct {
-		Diff []interface{} `json:"diff"`
+func (api *API) blueprintsDiffHandler(writer http.ResponseWriter, request *http.Request, params httprouter.Params) {
+	type pack struct {
+		Package blueprintPackage `json:"Package"`
 	}
 
-	reply.Diff = make([]interface{}, 0)
-	json.NewEncoder(writer).Encode(reply)
+	type diff struct {
+		New *pack `json:"new"`
+		Old *pack `json:"old"`
+	}
+
+	type reply struct {
+		Diffs []diff `json:"diff"`
+	}
+
+	name := params.ByName("blueprint")
+	if len(name) == 0 {
+		statusResponseError(writer, http.StatusNotFound, "no blueprint name given")
+		return
+	}
+	fromCommit := params.ByName("from")
+	if len(fromCommit) == 0 {
+		statusResponseError(writer, http.StatusNotFound, "no from commit ID given")
+		return
+	}
+	toCommit := params.ByName("to")
+	if len(toCommit) == 0 {
+		statusResponseError(writer, http.StatusNotFound, "no to commit ID given")
+		return
+	}
+
+	var oldBlueprint blueprint
+	if fromCommit == "NEWEST" {
+		if !api.store.getBlueprintCommitted(name, &oldBlueprint) {
+			statusResponseError(writer, http.StatusNotFound)
+			return
+		}
+	} else {
+		statusResponseError(writer, http.StatusNotFound, "invalid from commit ID given")
+		return
+	}
+
+	var newBlueprint blueprint
+	if toCommit == "WORKSPACE" {
+		if !api.store.getBlueprint(name, &newBlueprint, nil) {
+			statusResponseError(writer, http.StatusNotFound)
+			return
+		}
+	} else {
+		statusResponseError(writer, http.StatusNotFound, "invalid to commit ID given")
+		return
+	}
+
+	diffs := []diff{}
+
+	for _, oldPack := range oldBlueprint.Packages {
+		if !packageListContains(newBlueprint.Packages, oldPack) {
```

Comment:
This is O(n²) :cry: 

I'd copy one of the blueprints into a map, with names as keys, and the other one to a list. Then, iterate over the list  and for each package, check if it's in the map. If it is, delete it from both. At then end, the map will contain all added packages and the list all removed packages.

Reply:
I updated this per your advice except instead of dropping from the list I just added those packages directly to the diffs slice. I believe it runs in O(nlogn) now.